aospy/automate.py

```diff
@@ -0,0 +1,263 @@
+
+    Parameters
+    ----------
+    type_ : The desired type
+    parent : The object from which to get the attributes with type matching
+        'type_'
+
+    Returns
+    -------
+    A set (possibly empty) of attributes from 'parent'
+    """
+    return set([obj for obj in parent.__dict__.values()
+                if isinstance(obj, type_)])
+
+
+class CalcSuite(object):
+    """Suite of Calc objects generated from provided specifications."""
+
+    _CORE_SPEC_NAMES = {_OBJ_LIB_STR, _PROJECTS_STR, _MODELS_STR, _RUNS_STR}
+    _AUX_SPEC_NAMES = {_VARIABLES_STR,
+                       _REGIONS_STR,
+                       'date_ranges',
+                       'input_time_intervals',
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions'}
+    _NAMES_SUITE_TO_CALC = {
+        _PROJECTS_STR: 'proj',
+        _MODELS_STR: 'model',
+        _RUNS_STR: 'run',
+        _VARIABLES_STR: 'var',
+        _REGIONS_STR: 'region',
+        'date_ranges': 'date_range',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs):
+        self._specs_in = calc_suite_specs
+        self._obj_lib = self._specs_in[_OBJ_LIB_STR]
+
+    def _get_requested_spec(self, obj, spec_name):
+        """Helper to translate user specifications to needed objects."""
+        requested = self._specs_in[spec_name]
+        if isinstance(requested, str):
+            return [_get_attr_by_tag(obj, requested, spec_name)]
+        else:
+            return requested
+
+    def _permute_core_specs(self):
+        """Generate all requested combinations of the core objects."""
+        obj_trees = []
+        projects = self._get_requested_spec(self._obj_lib, _PROJECTS_STR)
+        for project in projects:
+            models = self._get_requested_spec(project, _MODELS_STR)
+            for model in models:
+                runs = self._get_requested_spec(model, _RUNS_STR)
+                for run in runs:
+                    obj_trees.append({
+                        self._NAMES_SUITE_TO_CALC[_PROJECTS_STR]: project,
+                        self._NAMES_SUITE_TO_CALC[_MODELS_STR]: model,
+                        self._NAMES_SUITE_TO_CALC[_RUNS_STR]: run,
+                    })
+        return obj_trees
+
+    def _get_regions(self):
+        """Get the requested regions."""
+        if self._specs_in[_REGIONS_STR] == 'all':
+            try:
```

Comment:
Instead of a try and except block, should we instead use ```getattr``` with a default argument?  

E.g. ```_get_all_objs_of_type(Region, getattr(self._obj_lib, 'regions', self._obj_lib))```